Fix message frame length parser

jwillmartin opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description

These fixes address issues with the existing data parsing.
Previous logic was inadvertently parsing for the embedded message length twice and ran into issues with creating incorrect vectors to test. The first searching logic was removed and second logic was corrected, per the attached IEEE 1609.3 snippet. 
![1609-3-length-encoding](https://github.com/user-attachments/assets/2c59a6e0-dac4-4a38-95ae-0ffe3db4dedd)

## Related GitHub Issue

<!--- This project only accepts pull requests related to open GitHub issues or Jira Keys -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please DO NOT name partially fixed issues, instead open an issue specific to this fix -->
<!--- Please link to the issue here: -->

## Related Jira Key

<!-- e.g. CAR-123 -->
VS-36

## Motivation and Context

Received data was not being parsed correctly and too many warnings were being displayed on the screen.

## How Has This Been Tested?
Tested with 8 pcap files containing, BSM/SPAT/MAP/MOM. The files were from 4 different sources and sending methods. 
All valid messages within the files were correctly detected. 

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [X] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

jwillmartin committed:
modified:   v2x_ros_driver/config/params.yaml
	modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h
	modified:   v2x_ros_driver/src/v2x_radio_client.cpp
	modified:   v2x_ros_driver/src/v2x_ros_driver_node.cpp

jwillmartin committed:
modified:   Dockerfile
	modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h
	modified:   v2x_ros_driver/src/v2x_radio_client.cpp
	modified:   v2x_ros_driver/src/v2x_ros_driver_node.cpp

jwillmartin committed:
modified:   v2x_ros_driver/config/params.yaml

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -142,49 +142,25 @@ void V2XRadioClient::close() {
     for (size_t i = 0; i < entry.size() - 3; i++) { // leave 3 bytes after (for lsb of id, length byte 1, and either message body or length byte 2)
         // Generate the 16-bit message id from two bytes, skip if it isn't a valid one
         uint16_t msg_id = (static_cast<uint16_t>(entry[i]) << 8) | static_cast<uint16_t>(entry[i + 1]);
+        if (!IsValidMsgID(std::to_string(msg_id))) { continue; }
```
I just realized we call this `msg_id` here, but call it `dsrc_id` in the wave.json. The J2735 standard calls id `DSRCmsgid`, is there a standard convention to use, or should we just match the standard?

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -142,49 +142,25 @@ void V2XRadioClient::close() {
+            if (msg_vec.size() > 16383)
+            {
+                RCLCPP_WARN_STREAM(logger_, "V2XRadioClient::process() : discarding received message with length field longer than 16383.");
+                continue;
```
Should this be a break statement, since we don't need to continue the loop in this case?

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -142,49 +142,25 @@ void V2XRadioClient::close() {
         uint16_t msg_id = (static_cast<uint16_t>(entry[i]) << 8) | static_cast<uint16_t>(entry[i + 1]);
+        if (!IsValidMsgID(std::to_string(msg_id))) { continue; }
+
+        if ((i + short_frame_) < entry.size()) {
```
If this fails, we will loop 2 more times and then discard the message without logging. Could we update this to be something similar to:

```
if ((i + short_frame_) > entry.size()) {
     // Warning log for discarding message
     // break;
}
```

jwillmartin replied:
Didn't update, i < entry.size() -3 already prevents additional loops from happening

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -208,14 +184,14 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
 {
-    // If message vector is larger than 255 bytes, length field will be 2 octets.
-    if (msg_vec.size() > 255) 
+    // If message vector is 255 bytes (128 octets) or larger, length field will be 2 octets.
```
I thought bytes and octets were approximately the same, is "255 bytes (128 octets)" accurate? Or can we remove the reference to bytes altogether?

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -225,11 +201,11 @@ bool V2XRadioClient::isValidMsgSize(const std::vector<uint8_t> msg_vec, size_t s
     }
-    // If message vector is smaller than 255 bytes, length field will be 1 octet. Additional check to make sure msg_size[2] exists.
-    else if (msg_vec.size() < 255 && msg_vec.size() >= 3)
+    // If message vector is smaller than 255 bytes (128 octets), length field will be 1 octet. Additional check to make sure msg_size[2] exists.
```
Similar to a previous comment, I thought bytes and octets were approximately the same, is "255 bytes (128 octets)" accurate? Or can we remove the reference to bytes altogether?

JonSmet commented on v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h:
```diff
@@ -162,7 +161,7 @@ class V2XRadioClient
     /** @brief WAVE Service Advertisement (WSA) frame size (bytes) for J2735 payloads >127 octets. IEEE 1609.3 (2020) - 8.1.3 */
     static const int long_frame_ = 4;
-    /** @brief WAVE Service Advertisement (WSA) frame size (bytes) for J2735 payloads <127 octets. IEEE 1609.3 (2020) - 8.1.3 */
+    /** @brief WAVE Service Advertisement (WSA) frame size (bytes) for J2735 payloads <128 octets. IEEE 1609.3 (2020) - 8.1.3 */
```
Where is it documented that WSA frame size is 4 for J2735 payloads >127 octets and 3 for J2735 payloads <128 octets?

JonSmet replied:
Let's update the comment to note that these include the two bytes for DSRCmsgID

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -208,14 +184,14 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
 }
 
-bool V2XRadioClient::isValidMsgSize(const std::vector<uint8_t> msg_vec, size_t start_index, size_t end_index, const std::vector<uint8_t> entry)
+bool V2XRadioClient::isValidMsgSize(const std::vector<uint8_t> msg_vec, size_t start_index, const std::vector<uint8_t> entry)
```
Missed this in the previous PR, but can we update isValidMsgSiz() to pass `msg_vec` and `entry` by const reference?

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -208,14 +184,14 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
-        std::vector<uint8_t> long_vec(entry.begin() + tmp_start_index, entry.begin() + end_index);
-        auto msg_size = (static_cast<uint16_t>(msg_vec[2]) << 8) | msg_vec[3];
+        std::vector<uint8_t> long_vec(entry.begin() + tmp_start_index, entry.end());
+        auto msg_size = (static_cast<uint16_t>(msg_vec[2] & 0x7F) << 8) | msg_vec[3];
```
It's a little unclear what is going on here, could we add a comment explaining?

jwillmartin committed:
modified:   v2x_ros_driver/config/wave.json
	modified:   v2x_ros_driver/etc/wave.json
	modified:   v2x_ros_driver/etc/wave_schema.json
	modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h
	modified:   v2x_ros_driver/include/v2x_ros_driver/v2x_ros_driver_node.h
	modified:   v2x_ros_driver/src/v2x_radio_client.cpp
	modified:   v2x_ros_driver/src/v2x_ros_driver_node.cpp

jwillmartin committed:
modified:   v2x_ros_driver/src/v2x_radio_client.cpp

JonSmet commented on v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h:
```diff
@@ -110,13 +110,13 @@ class V2XRadioClient
 
     /**
-     * @brief Validate possible Message ID
+     * @brief Validate possible DSRCmsgID
```
Let's update the name of this function from `IsValidMsgID()` to `isValidDSRCmsgID()`, if you agree?

Also consider changing the name of the input parameter from `msg_id` to `dsrc_msg_id`

jwillmartin replied:
Also didn't update to avoid confusion. Think it's fine?

JonSmet commented on v2x_ros_driver/include/v2x_ros_driver/v2x_radio_client.h:
```diff
@@ -131,16 +131,17 @@ class V2XRadioClient
      * @param msg_id Message ID under test
      */
-    bool isValidPSID(const std::string& msg_id);
+    bool isValidPSID(const std::string &msg_id);
```
Consider changing the input parameter from `msg_id` to `dsrc_msg_id` (also would require an update to the parameter description comment above)

jwillmartin replied:
I didn't update this msg_id because it could be confusing when it gets compared with the dsrc_msg_id list 
if (msg_id == dsrc_msg_id)

JonSmet commented on v2x_ros_driver/src/v2x_radio_client.cpp:
```diff
@@ -139,52 +139,36 @@ void V2XRadioClient::close() {
-        // Generate the 16-bit message id from two bytes, skip if it isn't a valid one
+        // Generate a 16-bit message id from two bytes, e.g. [0 20 ...] = 0x0014, skip if it isn't a valid one
         uint16_t msg_id = (static_cast<uint16_t>(entry[i]) << 8) | static_cast<uint16_t>(entry[i + 1]);
+        if (!IsValidMsgID(std::to_string(msg_id))) { continue; }
```
Let's rename the `msg_id` variable in this function to `dsrc_msg_id` to stay consistent with other usages in this package, if you agree?

jwillmartin replied:
Same as other comments

jwillmartin committed:
modified:   v2x_ros_driver/test/node_test.cpp

jwillmartin committed:
modified:   v2x_ros_driver/test/node_test.cpp
